README.md
# CLI

[![Build Status](https://travis-ci.org/Prowect/CLI.svg)](https://travis-ci.org/Prowect/CLI)
[![Code Climate](https://codeclimate.com/github/Prowect/CLI/badges/gpa.svg)](https://codeclimate.com/github/Prowect/CLI)
[![Test Coverage](https://codeclimate.com/github/Prowect/CLI/badges/coverage.svg)](https://codeclimate.com/github/Prowect/CLI/coverage)
[![Latest Release](https://img.shields.io/packagist/v/drips/cli.svg)](https://packagist.org/packages/drips/cli)

## Beschreibung

Dieses Library beinhaltet Klassen zur Interaktion mit der Konsole über PHP.

## Installation

1. Zuerst müssen alle Klassen des src-Verzeichnisses included werden. Dann sollten die Klassen unter der Verwendung des richtigen Namespaces verfügbar sein.

## Verwendung

### Console

#### Ausgabe auf der Konsole

Die Ausgabe auf der Konsole kann auf verschiedene Arten erfolgen. Die einfachste Möglichkeit ist einfach die Methode `writeln` zu verwenden.
Hierbei wird automatisch ein Zeilenumbruch an das Ende der Zeile angehängt.

```php
<?php
Console::writeln("Hello World");
```

Möchte man auf den Zeilenumbruch verzichten, so muss man die Methode `write` verwenden.

```php
<?php
Console::write("Hello World");
```

Außerdem können Fehler- und Erfolgsmeldungen generiert werden, welche farblich entsprechend hervorgehoben werden.

```php
<?php
Console::error("This is an error message");
Console::success("This is an success message");
```

#### Farben ändern

Die Schrift- und Hintergrundfarbe kann jederzeit geändert werden. Hierfür gibt es die Methoden `setColor` zum Ändern der Schriftfarbe und `setBgColor` zum Setzen der Hintergrundfarbe.

Mit der Methode `reset` werden die Farben zurückgesetzt.

```php
<?php
Console::setColor("blue");
Console::writeln("Blau");
Console::reset();
```

Folgende Farben stehen zur Verfügung:

**Schriftfarben:**

 - `black`
 - `dark_gray`
 - `blue`
 - `light_blue`
 - `green`
 - `light_green`
 - `cyan`
 - `light_cyan`
 - `red`
 - `light_red`
 - `purple`
 - `light_purple`
 - `brown`
 - `yellow`
 - `light_gray`
 - `white`
 - `black_u`
 - `red_u`
 - `green_u`
 - `yellow_u`
 - `blue_u`
 - `purple_u`
 - `cyan_u`
 - `white_u`

 **Hintergrundfarben**

 - `black`
 - `red`
 - `green`
 - `yellow`
 - `blue`
 - `magenta`
 - `cyan`
 - `light_gray`

### Kommandos

Außerdem gibt es die Möglichkeit benutzerdefinierte Kommandos für die Konsole zu erstellen.
Hierfür definiert man eine Klasse die das `ICommand`-Interface implementiert.

Anschließend kann das Kommando registriert werden.

```php
<?php
Command::register("mycmd", MyCommand::class);
```

Anschließend kann das Tool gestartet werden und versucht die übergebenen `$argv` auszuwerten.

```php
<?php
Command::execute($argv);
```

> `$argv` beinhaltet an erster Stelle den Namen der Datei, dieser wird sozusagen ignoriert. Das 2. Element wird als Kommando-Name aufgefasst unter welchem das Kommando registriert wurde. Das 3. Element stellt die Methode da, die von der Kommando-Klasse aufgerufen werden soll, während die restlichen Argumente als Parameter an die entsprechende Methode übergeben werden.

#### Beispiel

```php
<?php

abstract class MyCommand implements ICommand
{
    public static function exec($param = "")
    {
        Console::writeln("Führe exec mit $param aus.");
    }

    public static function help()
    {
        Console::writeln("Dies ist die Hilfe von MyCommand");
    }
}

Command::register("mycommand", MyCommand::class);
Command::execute($argv);
```

> In diesem Fall wird beim Aufruf des Scripts über die Konsole die `exec`-Methode der `MyCommand`-Klasse aufgerufen wenn das Script wie folgt aufgerufen wird. `php name_des_scripts.php mycommand exec 123` (123 wird als `$param` eingesetzt)


CHANGELOG.md
# Changelog

## Version 0.2

Dateien umbenannt


## Version 0.1

 `Composer PSR-4`


tests/ConsoleTest.php
<?php

namespace tests;

use Drips\CLI\Console;
use PHPUnit_Framework_TestCase;

class ConsoleTest extends PHPUnit_Framework_TestCase
{
    public function testConsoleOutput()
    {
        $str = "Hello World";
        ob_start();
        Console::write($str);
        $content = ob_get_contents();
        ob_end_clean();
        $this->assertEquals($str, $content);
    }

    public function testConsoleNlOutput()
    {
        $str = "Hello World";
        ob_start();
        Console::writeln($str);
        $content = ob_get_contents();
        ob_end_clean();
        $this->assertEquals($str.PHP_EOL, $content);
    }

    public function testBgColor()
    {
        Console::setBgColor("red");
        Console::writeln("Rot");
        Console::reset();
    }

    public function testSuccessOutput()
    {
        Console::success("Success");
    }

    public function testErrorOutput()
    {
        Console::error("Error");
    }
}


tests/TestCommand.php
<?php

namespace tests;

use Drips\CLI\ICommand;

require_once __DIR__.'/../src/console.php';
require_once __DIR__.'/../src/command.php';
require_once __DIR__.'/../src/icommand.php';

abstract class TestCommand implements ICommand
{
    public static function command1($param = "")
    {
        echo "TestCommand: $param";
    }

    public static function help()
    {
        echo "TestCommand: HELP";
    }
}


tests/TestConsole.php
<?php

namespace tests;

use Drips\CLI\Console;
use PHPUnit_Framework_TestCase;

require_once __DIR__.'/../vendor/autoload.php';

class TestConsole extends PHPUnit_Framework_TestCase
{
    public function testConsoleOutput()
    {
        $str = "Hello World";
        ob_start();
        Console::write($str);
        $content = ob_get_contents();
        ob_end_clean();
        $this->assertEquals($str, $content);
    }

    public function testConsoleNlOutput()
    {
        $str = "Hello World";
        ob_start();
        Console::writeln($str);
        $content = ob_get_contents();
        ob_end_clean();
        $this->assertEquals($str."\n\r", $content);
    }

    public function testBgColor()
    {
        Console::setBgColor("red");
        Console::writeln("Rot");
        Console::reset();
    }

    public function testSuccessOutput()
    {
        Console::success("Success");
    }

    public function testErrorOutput()
    {
        Console::error("Error");
    }
}


src/command.php
<?php

/**
 * Created by Prowect
 * Author: Raffael Kessler
 * Date: 30.10.2015 - 17:00
 * Copyright: Prowect
 */

namespace Drips\CLI;

use ReflectionClass;

/**
 * Class Command
 *
 * Wird verwendet um selbsterstellte Kommandos, die das ICommand-Interface
 * implementieren automatisiert auszuführen.
 * Hierfür müssen die Kommandos registriert werden.
 */
abstract class Command
{
    /**
     * Beinhaltet alle registrierten Commandos
     *
     * @var array
     */
    protected static $commands = array();

    /**
     * Dient zum Registrieren von Kommandos.
     * Gibt zurück ob die Registrierung erfolgreich war.
     * Wird FALSE zurückgegeben, ist das Kommando bereits registriert, bzw.
     * der Name des Kommandos bereits vergeben.
     *
     * @param string $command Name des Kommandos, unter welchem es erreichbar sein soll
     * @param ICommand $class Klasse die die Kommandos beinhaltet
     *
     * @return bool
     */
    public static function register($command, $class)
    {
        // Überprüft ob das Kommando auch das ICommand-Interface implementiert
        $reflection = new ReflectionClass($class);
        $icmd = new ReflectionClass(ICommand::class);
        $interfaces = $reflection->getInterfaces();
        if(!in_array($icmd, $interfaces)){
            return false;
        }
        $command = strtolower($command);
        if(!isset(static::$commands[$command])){
            static::$commands[$command] = $class;
            return true;
        }
        return false;
    }

    /**
     * Führt automatisch das entsprechende Kommando aus, bzw. erzeugt eine
     * Fehlermeldung wenn ein Kommando nicht richtig aufgerufen wurde.
     * Gibt TRUE/FALSE zurück, je nachdem ob übergebene Methode der Klasse
     * aufgerufen werden konnte oder nicht.
     *
     * @param array $args Args die von PHP übergeben wurden
     *
     * @return bool
     */
    public static function execute(array $args)
    {
        if(!empty($args)){
            $command = strtolower($args[1]);
            if(!isset(static::$commands[$command])){
                Console::error("Command ($command) not found!");
            } else {
                $commandClass = static::$commands[$command];
                if(count($args) >= 3){
                    $method = $args[2];
                    if(method_exists($commandClass, $method)){
                        $args = array_slice($args, 3);
                        call_user_func_array(array($commandClass, $method), $args);
                        return true;
                    }
                }
                $commandClass::help();
            }
        }
        return false;
    }

    /**
     * Entfernt ein registriertes Kommando
     * 
     * @param string $command Name des Kommandos, das entfernt werden soll
     */
    public static function unregister($command)
    {
        if(isset(static::$commands[$command])){
            unset(static::$commands[$command]);
        }
    }

    /**
     * Gibt eine Liste mit allen registrierten Kommandos und deren Funktionen aus.
     */
    public static function help()
    {
        Console::writeln();
        foreach(static::$commands as $command => $class){
            Console::writeln("[$command]");
            foreach(get_class_methods($class) as $method){
                Console::writeln(" - $method");
            }
        }
        Console::writeln();
    }
}


src/console.php
<?php

/**
 * Created by Prowect
 * Author: Raffael Kessler
 * Date: 30.10.2015 - 16:30
 * Copyright: Prowect
 */

namespace Drips\CLI;

/**
 * Class Console
 *
 * Dient als Schnittstelle zwischen PHP und der Konsole.
 * Kann zum Einlesen und Ausgeben auf die Konsole verwendet werden.
 */
abstract class Console
{
    /**
     * Beinhaltet die Farbcodes der Schriftfarben
     *
     * @var array
     */
    private static $fgColors = array(
       'black' => '0;30',
       'dark_gray' => '1;30',
       'blue' => '0;34',
       'light_blue' => '1;34',
       'green' => '0;32',
       'light_green' => '1;32',
       'cyan' => '0;36',
       'light_cyan' => '1;36',
       'red' => '0;31',
       'light_red' => '1;31',
       'purple' => '0;35',
       'light_purple' => '1;35',
       'brown' => '0;33',
       'yellow' => '1;33',
       'light_gray' => '0;37',
       'white' => '1;37',
       'black_u' => '4;30',
       'red_u' => '4;31',
       'green_u' => '4;32',
       'yellow_u' => '4;33',
       'blue_u' => '4;34',
       'purple_u' => '4;35',
       'cyan_u' => '4;36',
       'white_u' => '4;37',
   );

   /**
    * Beinhaltet die Farbcodes der Hintergrundfarben
    *
    * @var array
    */
   private static $bgColors = array(
       'black' => '40',
       'red' => '41',
       'green' => '42',
       'yellow' => '43',
       'blue' => '44',
       'magenta' => '45',
       'cyan' => '46',
       'light_gray' => '47',
   );

   /**
    * Setzt die aktuelle Farbe (auf die Standardfarbe) zurück
    */
   public static function reset()
   {
       echo "\033[0m";
   }

   /**
    * Setzt die Schriftfarbe auf die übergebene Farbe. Gültige Farben befinden
    * sich in der $fgColors
    *
    * @param string $color Die Schriftfarbe die verwendet werden soll
    */
   public static function setColor($color)
   {
       if (isset(self::$fgColors[$color])) {
           echo "\033[".self::$fgColors[$color].'m';
       }
   }

   /**
    * Setzt die Hintergrundfarbe die verwendet werden soll. Gültige Farben befinden
    * sich in der $bgColors
    *
    * @param string $color Die Hintergrundfarbe die verwendet werden soll
    */
   public static function setBgColor($color)
   {
       if (isset(self::$bgColors[$color])) {
           echo "\033[".self::$bgColors[$color].'m';
       }
   }

   /**
    * Gibt einen übergebenen String mit einem echo auf der Konsole aus.
    *
    * @param string $str Der String der ausgegeben werden soll
    * @param bool $newLine Erzeugt einen Zeilenumbruch wenn TRUE
    */
   public static function write($str = "", $newLine = false)
   {
       echo $str;
       echo $newLine ? "\n\r" : '';
   }

   /**
    * Gibt einen übergebenen String mit einem echo auf der Konsole aus.
    * Ruft prinzipiell die write-Methode auf und erzeugt zusätzlich einen
    * Zeilenumbruch.
    *
    * @param  [type] $str [description]
    */
   public static function writeln($str = "")
   {
       self::write($str, true);
   }

   /**
    * Liest von der Konsole
    *
    * @return string
    */
   public static function read()
   {
       $stream = fopen('php://stdin', 'r');

       return fgets($stream);
   }

   /**
    * Gibt einen Fehler auf der Konsole aus. Hierbei wird lediglich ein writeln
    * durchgeführt, vordem die Schriftfarbe rot gesetzt wird. Anschließend wird
    * die Farbe wieder zurückgesetzt!
    *
    * @param string $str Die Fehlermeldung die ausgegeben werden soll
    */
   public static function error($str)
   {
       self::setColor('red');
       self::writeln($str);
       self::reset();
   }

   /**
    * Gibt eine Erfolgsmeldung auf der Konsole aus. Hierbei wird lediglich ein
    * writeln durchgeführt, vordem die Schriftfarbe grün gesetzt wird.
    * Anschließend wird die Farbe wieder zurückgesetzt.
    *
    * @param string $str Die Erfolgsmeldung die ausgegeben werden soll
    */
   public static function success($str)
   {
       self::setColor('green');
       self::writeln($str);
       self::reset();
   }
}


src/Console.php
<?php

/**
 * Created by Prowect
 * Author: Raffael Kessler
 * Date: 30.10.2015 - 16:30
 * Copyright: Prowect
 */

namespace Drips\CLI;

/**
 * Class Console
 *
 * Dient als Schnittstelle zwischen PHP und der Konsole.
 * Kann zum Einlesen und Ausgeben auf die Konsole verwendet werden.
 */
abstract class Console
{
    /**
     * Beinhaltet die Farbcodes der Schriftfarben
     *
     * @var array
     */
    private static $fgColors = array(
       'black' => '0;30',
       'dark_gray' => '1;30',
       'blue' => '0;34',
       'light_blue' => '1;34',
       'green' => '0;32',
       'light_green' => '1;32',
       'cyan' => '0;36',
       'light_cyan' => '1;36',
       'red' => '0;31',
       'light_red' => '1;31',
       'purple' => '0;35',
       'light_purple' => '1;35',
       'brown' => '0;33',
       'yellow' => '1;33',
       'light_gray' => '0;37',
       'white' => '1;37',
       'black_u' => '4;30',
       'red_u' => '4;31',
       'green_u' => '4;32',
       'yellow_u' => '4;33',
       'blue_u' => '4;34',
       'purple_u' => '4;35',
       'cyan_u' => '4;36',
       'white_u' => '4;37',
   );

   /**
    * Beinhaltet die Farbcodes der Hintergrundfarben
    *
    * @var array
    */
   private static $bgColors = array(
       'black' => '40',
       'red' => '41',
       'green' => '42',
       'yellow' => '43',
       'blue' => '44',
       'magenta' => '45',
       'cyan' => '46',
       'light_gray' => '47',
   );

   /**
    * Setzt die aktuelle Farbe (auf die Standardfarbe) zurück
    */
   public static function reset()
   {
       echo "\033[0m";
   }

   /**
    * Setzt die Schriftfarbe auf die übergebene Farbe. Gültige Farben befinden
    * sich in der $fgColors
    *
    * @param string $color Die Schriftfarbe die verwendet werden soll
    */
   public static function setColor($color)
   {
       if (isset(self::$fgColors[$color])) {
           echo "\033[".self::$fgColors[$color].'m';
       }
   }

   /**
    * Setzt die Hintergrundfarbe die verwendet werden soll. Gültige Farben befinden
    * sich in der $bgColors
    *
    * @param string $color Die Hintergrundfarbe die verwendet werden soll
    */
   public static function setBgColor($color)
   {
       if (isset(self::$bgColors[$color])) {
           echo "\033[".self::$bgColors[$color].'m';
       }
   }

   /**
    * Gibt einen übergebenen String mit einem echo auf der Konsole aus.
    *
    * @param string $str Der String der ausgegeben werden soll
    * @param bool $newLine Erzeugt einen Zeilenumbruch wenn TRUE
    */
   public static function write($str = "", $newLine = false)
   {
       echo $str;
       echo $newLine ? PHP_EOL : '';
   }

   /**
    * Gibt einen übergebenen String mit einem echo auf der Konsole aus.
    * Ruft prinzipiell die write-Methode auf und erzeugt zusätzlich einen
    * Zeilenumbruch.
    *
    * @param  [type] $str [description]
    */
   public static function writeln($str = "")
   {
       self::write($str, true);
   }

   /**
    * Liest von der Konsole
    *
    * @return string
    */
   public static function read()
   {
       $stream = fopen('php://stdin', 'r');

       return fgets($stream);
   }

   /**
    * Gibt einen Fehler auf der Konsole aus. Hierbei wird lediglich ein writeln
    * durchgeführt, vordem die Schriftfarbe rot gesetzt wird. Anschließend wird
    * die Farbe wieder zurückgesetzt!
    *
    * @param string $str Die Fehlermeldung die ausgegeben werden soll
    */
   public static function error($str)
   {
       self::setColor('red');
       self::writeln($str);
       self::reset();
   }

   /**
    * Gibt eine Erfolgsmeldung auf der Konsole aus. Hierbei wird lediglich ein
    * writeln durchgeführt, vordem die Schriftfarbe grün gesetzt wird.
    * Anschließend wird die Farbe wieder zurückgesetzt.
    *
    * @param string $str Die Erfolgsmeldung die ausgegeben werden soll
    */
   public static function success($str)
   {
       self::setColor('green');
       self::writeln($str);
       self::reset();
   }
}
